packages/interface-adapter/lib/adapter/web3/index.ts

```diff
@@ -49,8 +49,20 @@ export class Web3InterfaceAdapter implements InterfaceAdapter {
     return this.web3.eth.getAccounts();
   }
 
-  public estimateGas(transactionConfig: Transaction) {
-    return this.web3.eth.estimateGas(transactionConfig);
+  public async estimateGas(transactionConfig: Transaction, stacktrace = false) {
+    // web3 does not error gracefully when gas estimation fails due to a revert,
+    // so in cases where we want to get past this (debugging/stacktracing), we must
+    // catch the error and return null instead
+    if(stacktrace === true) {
+      try {
+        const gasEstimate = await this.web3.eth.estimateGas(transactionConfig);
+        return gasEstimate;
+      } catch {
```

Comment:
I have the same comment here, we should probably ensure that this is failing in whatever way `estimateGas` fails when it the tx always reverts. If it fails for another reason we should throw.

Reply:
In general I agree with this sentiment, but this is a very odd situation as far as how gas estimation errors are handled. It does not appear that geth (and thus web3) provides information that can properly allow us to distinguish between different error cases when calling `eth_estimateGas`. That is the reason for the try/catch, because we want to be able to skip the error here and get to a more informative error for someone attempting to debug. Does that make sense?

I think it would be great to have better error handling on the geth side, that we could then consume. I will be writing up an issue in that repo soon to that effect in hopes a change can be made (though a similar problem was brought up 4 years ago to no avail -- https://github.com/ethereum/go-ethereum/issues/15671).

Reply:
So I'm understanding that you are saying that when `estimateGas` fails for any reason, we cannot know right away why it is failing? It might be failing because the tx always reverts or it might be failing because the input is improperly formatted, we have no way of telling for sure why it is failing?

Because if, for example, the input is improper - then we want to throw an error here for sure and not tell the user something like "your tx always reverts". That would be very unhelpful. But if we can't tell for sure why and we just have to display a more general error message, then I guess that's what we have to do. :)

Reply:
Yeah unfortunately we won't know. I am more ok with this because we will still error here in almost all situations, basically unless someone is debugging.

Reply:
There is even a PR that got closed without being merged that doesn't directly address this but does mention that error handling for this call isn't great: https://github.com/ethereum/EIPs/pull/2724

I will write up an issue in the geth repo, but my sense is that the best way to advocate there for this is to ask for a separate endpoint, `eth_estimateGasVerbose` or something. I'll be trying to advocate there but in the meantime I'm afraid this is what we have to work with.